File: Cpp/0383_ransom_note.cpp
// https://leetcode.cn/problems/ransom-note/

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        int n_ran = ransomNote.size(), n_mag = magazine.size();
        if (n_ran > n_mag) {
            return false;
        }

        vector<int> cnt_mag(26, 0);
        for (char c: magazine) {
            cnt_mag[c - 'a'] = cnt_mag[c - 'a'] + 1;
        }
        for (char c: ransomNote) {
            cnt_mag[c - 'a'] = cnt_mag[c - 'a'] - 1;
            if (cnt_mag[c - 'a'] < 0) {
                return false;
            }
        }
        return true;
    }
};

File: Python3/0383_ransom_note.py
# https://leetcode.cn/problems/ransom-note/

class Solution:
    def canConstruct(self, ransomNote, magazine):
        """
        Inputs:
            ransomNote: str
            magzine: str
        Outputs:
            res: bool
        """

        if len(ransomNote) > len(magazine):
            return False

        cnt = [0 for _ in range(26)]
        for ch in magazine:
            cnt[ord(ch)-ord('a')] = cnt[ord(ch)-ord('a')] + 1
        for ch in ransomNote:
            cnt[ord(ch)-ord('a')] = cnt[ord(ch)-ord('a')] - 1
            if cnt[ord(ch)-ord('a')] < 0:
                return False
        return True